### PR TITLE
Augment hotels with regular price

### PR DESCRIPTION
**Pull request by votsa**
## Clubhouse Story

https://app.clubhouse.io/findhotel/story/34453/spike-expose-deal-score-in-sapi

[ch34453]

### CODE REVIEW

File: packages/core/src/search.ts [hunk 1]
```diff
@@ -206,6 +208,24 @@ function hotelsHaveStaticPosition(
   return true
 }
 
+function calculateDealScores(state: SearchResults) {
+  return {'12313': 0.5}
+}
+
+function augmentHotelWithRegularPrice(
```
Comment (jopdeklein) [inline packages/core/src/search.ts:215]: Should we call this `regularHistoricalPriceRange`? Just regular could be mixed up with something like prices with taxes etc. And range because it's a range :D

File: packages/core/src/search.ts [hunk 2]
```diff
@@ -391,19 +422,19 @@ export function search(base: Base): SearchFn {
     /** END */
 
     return {
-      loadOffers: async (objectId) => {
-        if (!objectId) {
+      loadOffers: async (hotelId) => {
```
Comment (jopdeklein) [inline packages/core/src/search.ts:425]: Nice :)

File: packages/core/src/utils/omit.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import {Except} from 'type-fest'
```
Comment (votsa) [inline packages/core/src/utils/omit.ts:1]: Just moved it to separate file

File: packages/core/src/utils/omit.test.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import {omit} from './omit'
```
Comment (votsa) [inline packages/core/src/utils/omit.test.ts:1]: Just moved it to separate file

File: packages/core/src/utils/search-id.test.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import {generateSearchId} from './search-id'
```
Comment (votsa) [inline packages/core/src/utils/search-id.test.ts:1]: Just moved it to separate file

File: packages/core/src/utils/search-id.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import hash from 'object-hash'
```
Comment (votsa) [inline packages/core/src/utils/search-id.ts:1]: Just moved it to separate file

File: packages/core/src/utils/state.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import {produce, Draft} from 'immer'
```
Comment (votsa) [inline packages/core/src/utils/state.ts:1]: Just moved it to separate file

File: packages/core/CHANGELOG.md [hunk 8]
```diff
@@ -14,6 +14,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 > - :house: [Internal]
 > - :nail_care: [Polish]
 
+## v0.0.15
```
Comment (jopdeklein) [inline packages/core/CHANGELOG.md:17]: ⚛️

File: packages/core/src/utils/deal-score.test.ts [hunk 9]
```diff
@@ -0,0 +1,220 @@
+import {
+  getDealScores,
+  calculateDealScores,
+  DealScoreParameters
+} from './deal-score'
+
+describe('getDealScores', () => {
+  it('returns empty deal scores object if no hotels', () => {
```
Comment (jopdeklein) [inline packages/core/src/utils/deal-score.test.ts:8]:
```suggestion
  it('returns empty deal scores object if there are no hotels', () => {
```

File: packages/core/src/utils/deal-score.test.ts [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+      hotelEntities: {},
+      hotelOfferEntities: {}
+    }
+    const expected = {}
+    expect(
+      getDealScores((dealScoreParameters as unknown) as DealScoreParameters)
+    ).toEqual(expected)
+  })
+
+  it('returns correct deal scores object when no available hotels', () => {
```
Comment (jopdeklein) [inline packages/core/src/utils/deal-score.test.ts:20]:
```suggestion
  it('returns correct deal scores object if there are no available hotels', () => {
```

File: packages/core/src/utils/deal-score.test.ts [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+      },
+      hotelOfferEntities: {}
+    }
+    const expected = {}
+    expect(
+      getDealScores((dealScoreParameters as unknown) as DealScoreParameters)
+    ).toEqual(expected)
+  })
+
+  it('returns correct deal scores object when there is only 1 hotel in results', () => {
```
Comment (jopdeklein) [inline packages/core/src/utils/deal-score.test.ts:35]:
```suggestion
  it('returns correct deal score object when there is only 1 hotel in results', () => {
```

File: packages/core/src/utils/deal-score.test.ts [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+        '4': {offers: [{nightlyRate: 150}, {id: '2', nightlyRate: 155}]}
+      }
+    }
+    const expected = {'1': 0.83, '2': 1, '3': 0}
+    expect(
+      getDealScores((dealScoreParameters as unknown) as DealScoreParameters)
+    ).toEqual(expected)
+  })
+
+  it('returns correct deal scores object when some hotels dont have historical price', () => {
```
Comment (jopdeklein) [inline packages/core/src/utils/deal-score.test.ts:178]: I think we should be consistent in our naming, should it be regular or historical price? (I like the latter more). I know this impacts the rest of the PR but it's easier to grasp IMO.

File: packages/core/src/utils/deal-score.ts [hunk 13]
```diff
@@ -4,7 +4,7 @@ import {getCheapestRate} from './cheapest-rate'
 
 import {HotelId, HotelOfferEntity, Hotel} from '../types'
 
-export type DealScores = Record<HotelId, number | undefined>
+export type DealScores = Record<HotelId, number>
```
Comment (velppa) [inline packages/core/src/utils/deal-score.ts:7]: I [learned](https://stackoverflow.com/questions/51936369/what-is-the-record-type-in-typescript) about Map type in TypeScript and in this case - unbounded values of key, HotelId; while Record implies some closed group of keys. Based on that it looks like Map should be used instead of Record in this case.


```typescript
Map<HotelId, number>
```

Or I understood it wrong?

File: packages/core/src/utils/deal-score.ts [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+ * @param input - DealScoreInput[]
+ *
+ * @returns DealScores
+ */
+export function calculateDealScores(input: DealScoreInput[] = []): DealScores {
+  const discounts: Record<HotelId, number> = {}
+  const dealScores: DealScores = {}
+
+  input.forEach((item) => {
+    const [hotelId, cheapestRate, regularFrom, regularTo] = item
```
Comment (jopdeklein) [inline packages/core/src/utils/deal-score.ts:28]:
```suggestion
    const [hotelId, cheapestRate, historicalLowerBound, historicalUpperBound] = item
```

I prefer lower / upper bound because the from / to implicates a "discount" in some way, like the "from" price.

Comment (jopdeklein) [inline packages/core/src/utils/deal-score.ts:28]: The typing helps, but I prefer to just spell these out in the function signature directly rather than the `DealScoreInput`. It's not quite clear which parts are optional or not and which can be omitted.

File: packages/core/src/utils/discount.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Calculate discount based on regular price range
+ *
+ * @param rate - price to calculate discount for
+ * @param from - regular price range from
+ * @param to - regular price range to
+ */
+export function calculateDiscount(rate: number, from?: number, to?: number) {
```
Comment (jopdeklein) [inline packages/core/src/utils/discount.ts:8]: Here I also prefer the lower  / upper bound naming over the from / to to avoid confusion (especially since this is dealing with a discount :))

File: packages/core/src/utils/discount.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+  if (from === undefined || to === undefined || to === 0) return 0
+
+  return rate < from
```
Comment (jopdeklein) [inline packages/core/src/utils/discount.ts:11]: For me the nested ternary operators are hard to parse, especially since the branches look so similar. Maybe in this case an if/if else/else is easier to grasp?

**Review by jopdeklein — CHANGES_REQUESTED** [review]

File: packages/core/src/utils/deal-score.ts [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  anchorHotelId?: HotelId
+  hotelIds: HotelId[]
+  hotelEntities: Record<HotelId, Hotel>
+  hotelOfferEntities: Record<HotelId, HotelOfferEntity>
+}
+
+/**
+ * Returns deal scores for provided hotels and offers
+ *
+ * @param parameters - parameters used to calculate deal score
```
Comment (jopdeklein) [inline packages/core/src/utils/deal-score.ts:69]: I believe we should expand this documentation with what a deal score is, or at least that it is a number between 0 and 1 (or whichever scale we do or will support).

File: packages/core/src/utils/deal-score.ts (hunk 13 above)
Comment (velppa) [reply packages/core/src/utils/deal-score.ts:7]: Update - turns out Map is hard to use in practice, it won't created out of object automatically, so I switched to Record back.
https://github.com/FindHotel/website-content/pull/12/commits/6ce045fe181905c0f7561b549a81e95302167274

File: packages/core/src/utils/deal-score.test.ts (hunk 12 above)
Comment (votsa) [reply packages/core/src/utils/deal-score.test.ts:178]: It's regular historical price. (we have regular for weekdays and prices for the weekends)

File: packages/core/src/utils/deal-score.ts (hunk 14 above)
Comment (votsa) [reply packages/core/src/utils/deal-score.ts:28]: But how we should do that? The function requires an array of values as we calculate deal score related to the list of the hotels, not for an individual hotel.

File: packages/core/src/utils/deal-score.test.ts (hunk 12 above)
Comment (jopdeklein) [reply packages/core/src/utils/deal-score.test.ts:178]: OK clear, should we then call it `regularHistorical` or `weekdayHistorical` ?

Comment (votsa) [reply packages/core/src/utils/deal-score.test.ts:178]: Not a `weekday` or `weekend` as we use `medianRateBkt` which covers both. I agree that the name then looks more verbose but it's quite long as well `regularHistoricalLowerBound` and `regularHistoricalUpperBound`.  which make the function bigger and adds more lines to the code. Wouldn't it be better to use shorter names but with description in the comment for the function?

File: packages/core/src/utils/deal-score.ts (hunk 14 above)
Comment (jopdeklein) [reply packages/core/src/utils/deal-score.ts:28]: I meant by defining the names in the type, it's a bit verbose but the intent is more clear:
```
[
  hotelId: HotelId,
  cheapestRate?: number,
  regularHistoricalBounds: [lowerBound: number, upperBound: number],
] 
```

Or by passing objects if you think's less verbose?

Comment (votsa) [reply packages/core/src/utils/deal-score.ts:28]: No, I don't like passing an object here. It will make the test function huge... Check [this test ](https://github.com/FindHotel/sapi/pull/34/files#diff-ad1cf1caf9f049f509be77dbb816d8d3559edc9c3064f180df2237c667b69b0fR201) and imagine that we should add key to each value... 
The named tuple type looks ok to me.

File: packages/core/src/utils/deal-score.ts (hunk 17 above)
Comment (votsa) [reply packages/core/src/utils/deal-score.ts:69]: Probably @velppa can help with the description?

**Review by jopdeklein — APPROVED** [review]
